editor/store/effects.js

```diff
@@ -154,14 +157,61 @@ export default {
 
 		request.then(
 			( newPost ) => {
-				const reset = isAutosave ? resetAutosave : resetPost;
-				dispatch( reset( newPost ) );
+				// An autosave may be processed by the server as a regular save
+				// when its update is requested by the author and the post was
+				// draft or auto-draft.
+				const isRevision = newPost.id !== post.id;
+
+				// Thus, the following behaviors occur:
+				//
+				//  - If it was a revision, it is treated as latest autosave
+				//    and updates optimistically applied are reverted.
+				//  - If it was an autosave but not revision under the above
+				//    noted conditions, cherry-pick updated properties since
+				//    the received revision entity shares some but not all
+				//    properties of a post.
+				//  - Otherwise, it was a full save and the received entity
+				//    can be considered the new reset post.
+				let updateAction;
+				if ( isRevision ) {
+					updateAction = resetAutosave( newPost );
+				} else if ( isAutosave ) {
+					const revisionEdits = pick( newPost, [
+						// Autosave content fields.
+						'title',
+						'content',
+						'excerpt',
+
+						// UI considers save to have occurred if modified date
+						// of post changes (e.g. PostPreviewButton).
+						//
+						// TODO: Consider formalized pattern for identifying a
+						// save as having completed.
+						'date',
+						'date_gmt',
+						'modified',
+						'modified_gmt',
+					] );
+
+					updateAction = updatePost( revisionEdits );
+				} else {
+					updateAction = resetPost( newPost );
+				}
+
+				dispatch( updateAction );
 
 				dispatch( {
 					type: 'REQUEST_POST_UPDATE_SUCCESS',
 					previousPost: post,
 					post: newPost,
-					optimist: { type: COMMIT, id: POST_UPDATE_TRANSACTION_ID },
+					optimist: {
+						// Note: REVERT is not a failure case here. Rather, it
+						// is simply reversing the assumption that the updates
+						// were applied to the post proper, such that the post
+						// treated as having unsaved changes.
+						type: isRevision ? REVERT : COMMIT,
```

Comment:
If the server returns a different excerpt (or title, content, etc... ) on the autosave endpoint ( because of a filter, for example) this revert is going to revert the changes made by the server, and they are not going to be reflected by UI right? This is something that I'm not sure if it is even possible but I may be missing something, and it may be helpful for me to understand the changes.